lib/scanner/walk.go

```diff
@@ -105,10 +113,77 @@ func (w *Walker) Walk() (chan protocol.FileInfo, error) {
 				filepath.Walk(filepath.Join(w.Dir, sub), hashFiles)
 			}
 		}
-		close(files)
+		close(toHashChan)
 	}()
 
-	return hashedFiles, nil
+	// We're not required to emit scan progress events, just kick off hashers,
+	// and feed inputs directly from the walker.
+	if w.ProgressTickIntervalS < 0 {
+		newParallelHasher(w.Dir, w.BlockSize, w.Hashers, finishedChan, toHashChan, nil, nil)
+		return finishedChan, nil
+	}
+
+	// Defaults to every 2 seconds.
+	if w.ProgressTickIntervalS == 0 {
+		w.ProgressTickIntervalS = 2
+	}
+
+	ticker := time.NewTicker(time.Duration(w.ProgressTickIntervalS) * time.Second)
+
+	// We need to emit progres events, hence we create a routine which buffers
+	// the list of files to be hashed, counts the total number of
+	// bytes to hash, and once no more files need to be hashed (chan get closed),
+	// start a routine which periodically emits FolderScanProgress events,
+	// until a stop signal is sent by the parallel hasher.
+	// Parallel hasher is stopped by this routine when we close the channel over
+	// which it receives the files we ask it to hash.
+	go func() {
```

Comment:
Do we need to provide updates more than once per file? If not, would it perhaps be cleaner to add more goroutines to the chain? E/goto:

FileInfos from the walker :arrow_right: Your buffering goroutine, keeps track of how much there is to scan :arrow_right: 
All the parallell hashers :arrow_right: Another goroutine that just intercepts the results from the hasher, bumps the progress counter according to that file's size and possibly generates the progress events :arrow_right: The usual database updater at the end of the chain...

The advantage of this is that we don't have to pass around a counter that needs atomic updates, we just need to calculate a total size and pass that to the interceptor goroutine when it starts...